### PR TITLE
[omnibus] 2.4.2, 2.4.3 -> 2.5

### PR DESCRIPTION
**Pull request by endofline**
**Compatibility Notice:** Upgraded plugins for this version of the bot are not 
  backward-compatible with any other version of the bot before ~~2.4.3~~ 2.5

* framework
  * adapted whitelist architecture proposed in #177, #182 (7ec1dac)
  * **reprocessor** for sending hidden context with bot messages - first seen as plugin in 
     da5a9ae, integrated as framework feature in #235
  * **convmem** as workaround for hangups "forgotten conversations" - first seen as plugin in
      7ab0a47, integrated as framework feature in #235 - stores conversation title, sparse users 
      list (chatid, gaiaid, full name) and conversation type based on [hangups.schemas.ConversationType](https://github.com/tdryer/hangups/blob/f06b19d12be07a3acce066b92ea3c35ac23024cd/hangups/schemas.py#L41) with intelligent guessing for existing memory entries
  * `bot.conversations.get(string convfilter)` gets filtered conversation dict, 
     call without arguments to get all conversations in memory - **deprecates
     `utils.get_all_conversations()`** - warning in log if used
  * `bot.conversations.get_name(<conv|convid>)` to get conversation name, 
     drop-in replacement for `hangups.ui.utils.get_conv_name` and `utils.get_conv_name`
     (**`utils.get_conv_name` is deprecated**, warning in log if used)
  * added **threadmanager** library to standardise thread management (#240)
  * added **webbridge** library to facilitate development of web-chat bridges (#240, 
    derived from #209, #201)
  * `/bot help` differentiates users with no one-to-one and users who are `/bot opt-out`.
     users who are `/bot optout` will be opted-in if they private message the bot (#242)
* new plugins
  * **example_reprocessor** to demonstrate *reprocessor* feature: `/bot testcontext`
  * **unittest_convmem** to demonstrate *convmem* feature: `/bot dumpconv`
  * **tldr** plugin for creating a TL;DR within a chat - documented on [plugin list]       
    (https://github.com/hangoutsbot/hangoutsbot/wiki/Plugin-List)
  * **example_webbridge** to demonstrate the *webbridge* framework library
* improved plugins: 
  * api
    * rewritten to use 2.4 pattern
    * security warning if GET method used
    * *reprocessor* nonce integration for better security
  * autoreply
    * fix regex metacharacters in keyword search (#192, ee1adb3)
      * prefix trigger phrases with `regex:` to prevent implicit re.escape 
    * reduce log verbosity when disabled
    * support list of random responses (#239)
  * convtools_invitations
    * refactored to use convmem feature (#241, 3c30a21)
    * when creating new groups, group will be renamed to `GROUP: <local date and time>`
      * groups only created when a valid non-zero invitee list is made (0efc8e5)
      * new group creation sleeps for a second to allow convmem title update (0efc8e5)
    * resolution for erratic behaviour reported in #193
    * behaviour of rsvp confirmed (re: #114)
    * better logging
  * default
    * restored `/bot user` and `/bot users` to plugin default (re: #186)
    * new convid-family commands (#224, [wiki](https://github.com/hangoutsbot/hangoutsbot/wiki/Default-Commands-Plugin#convid-family]))
    * `/bot hangouts` and `/bot broadcast` uses *convmem* to eliminate forgotten
       conversations bug due to incomplete `bot._conv_list`
    * "fix" for `/bot leave` in 1-to-1 conversations - memory no longer deleted, 
       appropriate warning logged (#235)
    * added `/bot config test` to allow botmins to test their json snippets (650d51e)
  * slack
    * user id and channel id resolution
    * *reprocessor* nonce integration for better security
    * detects and replaces the text for any slack-to-hangout chat text that begins with bot alias
       to `/<bot alias> [REDACTED]` to prevent slackers from running commands as bot
  * subscribe, mention
    * recognise mentions/keywords in messages relayed from slack
    * display actual source name (as `<slack username>@slack`)
  * mentions
    * `config.mentionminlength` to adjust minimum matched length 
      (previously hard-coded 2 chars) (re: #208,  414dfd7)
    * case-sensitive whole fragment matching added as an intermediary step
      between nickname match and standard match (re: #208, 006ab1a)
    * `/bot bemorespecific` toggle for bot's "be more specific" message on multiple
       users matched (re: #207, 0b26a5c)
* miscellaneous
  * docstrings for individual plugin commands 
    (3976ea9, a36d884, 4f382b1, e989be7, bb11181, d1d6e0f, 9d81a55)
  * locales split into its own [translation project]
    (https://github.com/hangoutsbot/hangoutsbot-locales) (re: #194)
  * set bot locale with `export HANGOUTSBOT_LOCALE=<language>` (baa9880)
  * removed non-working build files (re: #205)
  * `config.json` clean up (re: #216)
  * removed 0.2s sleep on handle_command, may reduce latency (3369170)
  * removed prints in favor of `logging` python standard library

### CODE REVIEW

**Comment by nylonee** [comment]
All framework changes, new plugins and plugin modifications should be documented on the wiki

**Comment by nylonee** [comment]
All commits for #217 up until this point have been documented in this PR (and accounted for in the wiki)
(Just using this comment as a placeholder)

**Comment by endofline** [comment]
Going to test this for another 24 hours, with any additional 2.4.3 blockers cleared, then we can merge to staging. There are a lot of drastic changes that will not be backward-compatible with older bots.

**Comment by endofline** [comment]
This specific merge has bug #244 - which was fixed in the subsequent release: https://github.com/hangoutsbot/hangoutsbot/releases/tag/2.5.0.patch-1
